packages/optimizer/lib/transformers/AutoExtensionImporter.js

```diff
@@ -0,0 +1,307 @@
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const {findMetaViewport} = require('../HtmlDomHelper');
+const {calculateHost} = require('../RuntimeHostHelper');
+const {AMP_FORMATS} = require('../AmpConstants');
+
+const BIND_SHORT_FORM_PREFIX = 'bind';
+const AMP_BIND_DATA_ATTRIBUTE_PREFIX = 'data-amp-bind-';
+const DEFAULT_FORMAT = 'AMP';
+
+// Some AMP component don't bring their own tag, but enable new attributes on other
+// elements. Most are included in the AMP validation rules, but some are not. These
+// need to be defined manually here.
+const manualAttributeToExtensionMapping = new Map([
+  ['mask', 'amp-inputmask'],
+  ['lightbox', 'amp-lightbox-gallery'],
+]);
+
+/**
+ * Extension Auto Importer - this transformer auto imports all missing AMP extensions.
+ *
+ * The importer analyzes the HTML source code and identifies missing AMP extension imports
+ * using multiple strategies:
+ *
+ * - use validation rules to map used AMP tags to required AMP extensions.
+ * - use validation rules to map used AMP attributes to required AMP extensions.
+ * - manually specifiy attribute to extension mappings if this information is not available in the
+ *   validation rules.
+ * - mnullay implement AMP extension detection for a few corner cases.
+ *
+ * This importer also enables a shortcode `bindtext` instead of `data-amp-bind-text` for specifying
+ * AMP bindings when the square bracket notation (`[text]`) is not available. To avoid accidently
+ * rewriting non-AMP attributes, the transformer uses the AMP validation rules to only rename bindable
+ * attributes as specified in the validation rules.
+ *
+ * This transformer supports the following option:
+ *
+ * - `format: [AMP|AMP4EMAIL|AMP4ADS]` - specifies the AMP format. Defaults to `AMP`.
+ * - `autoExtensionImport: [true|false]` - set to `false` to disable the auto extension import> Default to `true`.
+ */
+class AutoExtensionImporter {
+  constructor(config) {
+    this.enabled = config.autoExtensionImport !== false;
+    this.format = config.format || DEFAULT_FORMAT;
+    this.log_ = config.log.tag('AutoExtensionImporter');
+
+    // We use the validation rules to infer extension imports. The rules are downloaded once and for
+    // efficency, we initially extract all needed rules
+    this.initExtensionSpec_(config.validatorRules);
+  }
+
+  /**
+   * @private
+   */
+  async initExtensionSpec_(validatorRules) {
+    this.extensionSpec = validatorRules.fetch().then((rules) => {
+      // Map extension names to info required for generating the extension imports
+      const extensionsMap = new Map();
+      for (const ext of rules.extensions) {
+        if (ext.htmlFormat.includes(DEFAULT_FORMAT)) {
+          extensionsMap.set(ext.name, {
```

Comment:
This was actually a bug: `DEFAULT_FORMAT` needs to be `this.format`. Throwing an exception here feels like overkill, this would mean if the validation format has a bug it'll mean pages will suddenly start to break which we should avoid if possible.